In [30]:
#Load the Packages Needed
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings("ignore")

In [31]:
#Load the Data
pwt = pd.read_csv('https://raw.githubusercontent.com/jivizcaino/PWT_10.0/main/pwt100.csv', encoding="latin-1")
pwt

countrycode   country   currency_unit  year        rgdpe        rgdpo  \
0             ABW     Aruba  Aruban Guilder  1950          NaN          NaN   
1             ABW     Aruba  Aruban Guilder  1951          NaN          NaN   
2             ABW     Aruba  Aruban Guilder  1952          NaN          NaN   
3             ABW     Aruba  Aruban Guilder  1953          NaN          NaN   
4             ABW     Aruba  Aruban Guilder  1954          NaN          NaN   
...           ...       ...             ...   ...          ...          ...   
12805         ZWE  Zimbabwe       US Dollar  2015  40141.61719  39798.64453   
12806         ZWE  Zimbabwe       US Dollar  2016  41875.20313  40963.19141   
12807         ZWE  Zimbabwe       US Dollar  2017  44672.17578  44316.74219   
12808         ZWE  Zimbabwe       US Dollar  2018  44325.10938  43420.89844   
12809         ZWE  Zimbabwe       US Dollar  2019  42296.06250  40826.57031   

             pop       emp  avh        hc  ...     csh_x     csh_m     csh_r  \
0            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
1            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
2            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
3            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
4            NaN       NaN  NaN       NaN  ...       NaN       NaN       NaN   
...          ...       ...  ...       ...  ...       ...       ...       ...   
12805  13.814629  6.393752  NaN  2.584653  ...  0.140172 -0.287693 -0.051930   
12806  14.030331  6.504374  NaN  2.616257  ...  0.131920 -0.251232 -0.016258   
12807  14.236595  6.611773  NaN  2.648248  ...  0.126722 -0.202827 -0.039897   
12808  14.438802  6.714952  NaN  2.680630  ...  0.144485 -0.263658 -0.020791   
12809  14.645468  6.831017  NaN  2.713408  ...  0.213562 -0.270959 -0.089798   

           pl_c      pl_i      pl_g      pl_x      pl_m      pl_n      pl_k  
0           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
1           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
2           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
3           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
4           NaN       NaN       NaN       NaN       NaN       NaN       NaN  
...         ...       ...       ...       ...       ...       ...       ...  
12805  0.479228  0.651287  0.541446  0.616689  0.533235  0.422764  1.533909  
12806  0.470640  0.651027  0.539631  0.619789  0.519718  0.416510  1.491724  
12807  0.473560  0.639560  0.519956  0.619739  0.552042  0.415592  1.514525  
12808  0.543757  0.655473  0.529867  0.641361  0.561526  0.425143  1.590120  
12809  0.494755  0.652439  0.500927  0.487763  0.430082  0.420675  1.384068  

[12810 rows x 52 columns]

In [70]:
#Prepare the Dataset and Clean the Data
pwt_subset = pwt[['country','countrycode','year','cgdpo','pop','emp','avh','hc','cn','labsh','ctfp']]
#pwt_subset = pwt[['country','countrycode','year','cgdpo','pop','emp','hc','cn','labsh','ctfp']]
pwt_subset = pwt_subset[pwt_subset['year'] > 2009]
pwt_subset = pwt_subset.set_index('countrycode')
pwt_subset = pwt_subset.dropna()
pwt_subset

country  year        cgdpo        pop        emp  \
countrycode                                                          
ARG             Argentina  2010  768478.2500  40.895752  18.076746   
ARG             Argentina  2011  868765.0625  41.320500  18.488029   
ARG             Argentina  2012  882438.8125  41.755196  18.791292   
ARG             Argentina  2013  885112.9375  42.196030  19.124418   
ARG             Argentina  2014  864304.0000  42.637511  19.383051   
...                   ...   ...          ...        ...        ...   
ZAF          South Africa  2014  699348.0625  54.544186  16.992683   
ZAF          South Africa  2016  710312.5000  56.207646  17.984968   
ZAF          South Africa  2017  726021.0000  57.009756  18.302843   
ZAF          South Africa  2018  733319.3750  57.792518  18.686357   
ZAF          South Africa  2019  735067.0625  58.558270  18.642710   

                     avh        hc          cn     labsh      ctfp  
countrycode                                                         
ARG          1760.224782  2.830078  2876207.25  0.440692  0.777526  
ARG          1760.230369  2.858542  3099233.75  0.473124  0.828393  
ARG          1736.103360  2.887292  3156028.75  0.520179  0.819774  
ARG          1725.873553  2.916332  3166597.50  0.543918  0.813449  
ARG          1705.428368  2.945663  3204576.50  0.543918  0.781934  
...                  ...       ...         ...       ...       ...  
ZAF          2209.091467  2.675951  2727571.50  0.552341  0.596181  
ZAF          2218.455068  2.762902  2773613.50  0.568343  0.566500  
ZAF          2197.033263  2.809246  2791243.75  0.570414  0.570064  
ZAF          2191.363362  2.857641  2846327.75  0.570888  0.554365  
ZAF          2191.363362  2.908202  2897706.00  0.570888  0.542998  

[608 rows x 10 columns]

In [71]:
#Choose the Year to Perform the Analysis
print(pwt_subset.groupby("year").count())

      country  cgdpo  pop  emp  avh  hc  cn  labsh  ctfp
year                                                    
2010       61     61   61   61   61  61  61     61    61
2011       61     61   61   61   61  61  61     61    61
2012       61     61   61   61   61  61  61     61    61
2013       61     61   61   61   61  61  61     61    61
2014       60     60   60   60   60  60  60     60    60
2015       60     60   60   60   60  60  60     60    60
2016       61     61   61   61   61  61  61     61    61
2017       61     61   61   61   61  61  61     61    61
2018       61     61   61   61   61  61  61     61    61
2019       61     61   61   61   61  61  61     61    61


We will choose 2019: the latest year with the largets number of ebservations for our variables of interest.
*To Do: Robustness using 2010,2015, and an average for 2010-2019

In [72]:
#Descriptive Statistics for the Variables in our Dataset
print(pwt_subset[pwt_subset['year'] == 2019].describe([0.05,0.10,0.25,0.50,0.75,0.90,0.95]).round(2))

         year        cgdpo      pop     emp      avh     hc            cn  \
count    61.0        61.00    61.00   61.00    61.00  61.00  6.100000e+01   
mean   2019.0   1766051.39    86.23   40.07  1814.08   3.23  8.117018e+06   
std       0.0   3788319.57   252.13  120.07   240.46   0.43  1.636157e+07   
min    2019.0     17330.60     0.34    0.19  1380.61   2.17  6.244054e+04   
5%     2019.0     45079.35     0.87    0.46  1439.73   2.51  2.194441e+05   
10%    2019.0     71310.98     2.08    1.05  1505.56   2.70  3.189460e+05   
25%    2019.0    199241.38     5.53    2.85  1611.37   2.90  8.281802e+05   
50%    2019.0    496383.50    17.10    8.10  1791.74   3.27  2.566180e+06   
75%    2019.0   1862070.00    58.56   25.60  2019.92   3.59  8.721109e+06   
90%    2019.0   3120405.75   127.58   69.98  2142.10   3.72  1.885582e+07   
95%    2019.0   5031850.00   270.63  131.17  2168.92   3.77  2.610288e+07   
max    2019.0  20566034.00  1433.78  798.81  2330.17   4.35  1.015442e+08   

In [73]:
pwt_subset = pwt_subset[pwt_subset['year'] == 2019]
pwt_subset

country  year         cgdpo         pop         emp  \
countrycode                                                              
ARG              Argentina  2019  9.785588e+05   44.780677   20.643215   
AUS              Australia  2019  1.355143e+06   25.203198   12.863174   
AUT                Austria  2019  4.783293e+05    8.955102    4.550281   
BEL                Belgium  2019  5.236886e+05   11.539328    4.921937   
BGR               Bulgaria  2019  1.503254e+05    7.000119    3.420419   
...                    ...   ...           ...         ...         ...   
TUR                 Turkey  2019  2.246182e+06   83.429615   28.087334   
TWN                 Taiwan  2019  1.100433e+06   23.596027   11.500000   
URY                Uruguay  2019  7.131098e+04    3.461734    1.635379   
USA          United States  2019  2.056603e+07  329.064917  158.299591   
ZAF           South Africa  2019  7.350671e+05   58.558270   18.642710   

                     avh        hc            cn     labsh      ctfp  
countrycode                                                           
ARG          1609.068998  3.096804  3.374818e+06  0.543918  0.812184  
AUS          1726.797659  3.549666  5.899094e+06  0.571645  0.837903  
AUT          1611.374223  3.381046  2.875460e+06  0.583860  0.803706  
BEL          1586.430997  3.149034  3.496175e+06  0.594811  0.828750  
BGR          1645.246464  3.186015  4.582792e+05  0.527659  0.729627  
...                  ...       ...           ...       ...       ...  
TUR          1832.000000  2.514292  1.023572e+07  0.436831  0.876780  
TWN          2085.135333  3.357651  4.258740e+06  0.650921  0.819051  
URY          1532.513740  2.776406  3.189460e+05  0.473972  0.693409  
USA          1765.346390  3.749341  6.905909e+07  0.597091  1.000000  
ZAF          2191.363362  2.908202  2.897706e+06  0.570888  0.542998  

[61 rows x 10 columns]

In [74]:
pwt_subset['GDP_pw']       = pwt_subset['cgdpo']/pwt_subset['emp']
pwt_subset['GDP_pc']       = pwt_subset['cgdpo']/pwt_subset['pop']
pwt_subset['k']            = pwt_subset['cn']/pwt_subset['emp']
pwt_subset['emp_pop']      = pwt_subset['emp']/pwt_subset['pop']

pwt_subset['log_y_pw']       = np.log(pwt_subset['GDP_pw'])
pwt_subset['log_y_pc']       = np.log(pwt_subset['GDP_pc'])
pwt_subset['log_k']          = np.log(pwt_subset['k'])
pwt_subset['log_emp_pop']    = np.log(pwt_subset['emp_pop'])

pwt_subset['y_pw_relUS']     = pwt_subset['GDP_pw']/pwt_subset.at['USA','GDP_pw']
pwt_subset['y_pc_relUS']     = pwt_subset['GDP_pc']/pwt_subset.at['USA','GDP_pc']
pwt_subset['k_relUS']        = pwt_subset['k']/pwt_subset.at['USA','k']
pwt_subset['h_relUS']        = pwt_subset['hc']/pwt_subset.at['USA','hc']
pwt_subset['emp_pop_relUS']  = pwt_subset['emp_pop']/pwt_subset.at['USA','emp_pop']
pwt_subset

country  year         cgdpo         pop         emp  \
countrycode                                                              
ARG              Argentina  2019  9.785588e+05   44.780677   20.643215   
AUS              Australia  2019  1.355143e+06   25.203198   12.863174   
AUT                Austria  2019  4.783293e+05    8.955102    4.550281   
BEL                Belgium  2019  5.236886e+05   11.539328    4.921937   
BGR               Bulgaria  2019  1.503254e+05    7.000119    3.420419   
...                    ...   ...           ...         ...         ...   
TUR                 Turkey  2019  2.246182e+06   83.429615   28.087334   
TWN                 Taiwan  2019  1.100433e+06   23.596027   11.500000   
URY                Uruguay  2019  7.131098e+04    3.461734    1.635379   
USA          United States  2019  2.056603e+07  329.064917  158.299591   
ZAF           South Africa  2019  7.350671e+05   58.558270   18.642710   

                     avh        hc            cn     labsh      ctfp  ...  \
countrycode                                                           ...   
ARG          1609.068998  3.096804  3.374818e+06  0.543918  0.812184  ...   
AUS          1726.797659  3.549666  5.899094e+06  0.571645  0.837903  ...   
AUT          1611.374223  3.381046  2.875460e+06  0.583860  0.803706  ...   
BEL          1586.430997  3.149034  3.496175e+06  0.594811  0.828750  ...   
BGR          1645.246464  3.186015  4.582792e+05  0.527659  0.729627  ...   
...                  ...       ...           ...       ...       ...  ...   
TUR          1832.000000  2.514292  1.023572e+07  0.436831  0.876780  ...   
TWN          2085.135333  3.357651  4.258740e+06  0.650921  0.819051  ...   
URY          1532.513740  2.776406  3.189460e+05  0.473972  0.693409  ...   
USA          1765.346390  3.749341  6.905909e+07  0.597091  1.000000  ...   
ZAF          2191.363362  2.908202  2.897706e+06  0.570888  0.542998  ...   

              emp_pop   log_y_pw   log_y_pc      log_k  log_emp_pop  \
countrycode                                                           
ARG          0.460985  10.766449   9.992059  12.004465    -0.774390   
AUS          0.510379  11.565049  10.892447  13.035941    -0.672602   
AUT          0.508122  11.562866  10.885831  13.356534    -0.677034   
BEL          0.426536  11.574950  10.722891  13.473478    -0.852059   
BGR          0.488623  10.690795   9.974631  11.805471    -0.716164   
...               ...        ...        ...        ...          ...   
TUR          0.336659  11.289424  10.200739  12.806076    -1.088685   
TWN          0.487370  11.468867  10.750136  12.822137    -0.718731   
URY          0.472416  10.682931   9.933036  12.180903    -0.749895   
USA          0.481059  11.774662  11.042896  12.985984    -0.731766   
ZAF          0.318362  10.582262   9.437695  11.953975    -1.144567   

             y_pw_relUS  y_pc_relUS   k_relUS   h_relUS  emp_pop_relUS  
countrycode                                                             
ARG            0.364871    0.349645  0.374742  0.825960       0.958271  
AUS            0.810898    0.860321  1.051226  0.946744       1.060948  
AUT            0.809130    0.854648  1.448532  0.901771       1.056257  
BEL            0.818967    0.726145  1.628231  0.839890       0.886660  
BGR            0.338285    0.343604  0.307121  0.849753       1.015724  
...                 ...         ...       ...       ...            ...  
TUR            0.615550    0.430780  0.835347  0.670596       0.699829  
TWN            0.736537    0.746201  0.848872  0.895531       1.013120  
URY            0.335635    0.329605  0.447052  0.740505       0.982034  
USA            1.000000    1.000000  1.000000  1.000000       1.000000  
ZAF            0.303492    0.200849  0.356290  0.775657       0.661794  

[61 rows x 23 columns]

In [37]:
#An alternative way to compute variables is
#pwt_subset.assign( k = (pwt_subset.cn/pwt_subset.emp) )

# Development Accounting

In [103]:
#pwt_subset['y_hk_pw']     = (pwt_subset['k']**(1-pwt_subset['labsh']))*(( pwt_subset['hc']**pwt_subset['labsh']))
#pwt_subset['y_hk_pw']       = (pwt_subset['k']**(1/3))*(((pwt_subset['hc'])**(2/3) ))
pwt_subset['y_hk_pw']       = (pwt_subset['k']**(1/3))*(((pwt_subset['hc']*pwt_subset['avh'])**(2/3) ))
pwt_subset['y_hk_pw_relUS'] = pwt_subset['y_hk_pw']/pwt_subset.at['USA','y_hk_pw']
pwt_subset

pwt_subset.describe([0.05,0.10,0.25,0.50,0.75,0.90,0.95]).round(2)

year        cgdpo      pop     emp      avh     hc            cn  \
count    61.0        61.00    61.00   61.00    61.00  61.00  6.100000e+01   
mean   2019.0   1766051.39    86.23   40.07  1814.08   3.23  8.117018e+06   
std       0.0   3788319.57   252.13  120.07   240.46   0.43  1.636157e+07   
min    2019.0     17330.60     0.34    0.19  1380.61   2.17  6.244054e+04   
5%     2019.0     45079.35     0.87    0.46  1439.73   2.51  2.194441e+05   
10%    2019.0     71310.98     2.08    1.05  1505.56   2.70  3.189460e+05   
25%    2019.0    199241.38     5.53    2.85  1611.37   2.90  8.281802e+05   
50%    2019.0    496383.50    17.10    8.10  1791.74   3.27  2.566180e+06   
75%    2019.0   1862070.00    58.56   25.60  2019.92   3.59  8.721109e+06   
90%    2019.0   3120405.75   127.58   69.98  2142.10   3.72  1.885582e+07   
95%    2019.0   5031850.00   270.63  131.17  2168.92   3.77  2.610288e+07   
max    2019.0  20566034.00  1433.78  798.81  2330.17   4.35  1.015442e+08   

       labsh   ctfp     GDP_pw  ...  log_k  log_emp_pop  y_pw_relUS  \
count  61.00  61.00      61.00  ...  61.00        61.00       61.00   
mean    0.55   0.71   74177.78  ...  12.64        -0.74        0.57   
std     0.08   0.16   36303.48  ...   0.66         0.14        0.28   
min     0.32   0.40   18371.85  ...  11.08        -1.14        0.14   
5%      0.41   0.44   23559.62  ...  11.53        -0.96        0.18   
10%     0.44   0.51   31758.38  ...  11.75        -0.86        0.24   
25%     0.52   0.60   43949.43  ...  12.18        -0.82        0.34   
50%     0.56   0.72   71907.51  ...  12.81        -0.72        0.55   
75%     0.59   0.84   96217.70  ...  13.19        -0.66        0.74   
90%     0.64   0.91  106398.89  ...  13.35        -0.59        0.82   
95%     0.65   0.94  128644.07  ...  13.41        -0.57        0.99   
max     0.68   1.03  219595.21  ...  13.63        -0.29        1.69   

       y_pc_relUS  k_relUS  h_relUS  emp_pop_relUS   y_hk_pw  y_hk_pw_relUS  \
count       61.00    61.00    61.00          61.00     61.00          61.00   
mean         0.59     0.85     0.86           1.00  22349.48           0.84   
std          0.33     0.46     0.12           0.14   5310.23           0.20   
min          0.11     0.15     0.58           0.66  11467.78           0.43   
5%           0.18     0.23     0.67           0.80  13939.87           0.52   
10%          0.21     0.29     0.72           0.88  14741.15           0.55   
25%          0.34     0.45     0.77           0.91  18481.67           0.69   
50%          0.52     0.84     0.87           1.01  23448.25           0.88   
75%          0.75     1.23     0.96           1.07  26074.63           0.98   
90%          0.89     1.45     0.99           1.15  27759.20           1.04   
95%          1.20     1.52     1.01           1.18  29122.14           1.09   
max          1.63     1.89     1.16           1.56  38879.30           1.45   

       log_y_hk_pw  
count        61.00  
mean          9.98  
std           0.25  
min           9.35  
5%            9.54  
10%           9.60  
25%           9.82  
50%          10.06  
75%          10.17  
90%          10.23  
95%          10.28  
max          10.57  

[12 rows x 25 columns]

## GDP per worker

In [104]:
var_log_y_pw = np.var(pwt_subset['log_y_pw'])
round(var_log_y_pw,2)

0.28

In [105]:
var_log_y_pc = np.var(pwt_subset['log_y_pc'])
round(var_log_y_pc,2)

0.35

In [106]:
pwt_subset['log_y_hk_pw']  = np.log(pwt_subset['y_hk_pw'])
var_log_y_hk_pw            = np.var(pwt_subset['log_y_hk_pw'])
round(var_log_y_hk_pw,2)

0.06

In [107]:
var_log_tfp                = np.var(np.log(pwt_subset['ctfp']))
round(var_log_tfp,2)

0.06

In [108]:
success_factors = var_log_y_hk_pw/var_log_y_pw
round(success_factors,2)

0.22

In [109]:
success_TFP = var_log_tfp/var_log_y_pw
round(success_TFP,2)

0.2

In [110]:
round((success_factors + success_TFP),2)

0.42

**Conclusion 1:**

* Factors of production alone can explain 28% of the variation in income per worker
* TFP alone can explain 18% of the variation in income per worker
* More than half of the variation in income per worker remains unexplained
* A full variance decomposition needs to be performed to find out the determinants of the unexplained fraction of the variance

**Conclusion 1 (SE without avh):**

* I agree with Conclusion 1.

**Conclusion 1 (SE with avh):**

* Factors of production alone explain just 22%. 
* TFP alone explains 20% of the variation in income per worker.
* More than half (42% compared to 46% in the case without avh) of the variation in income per worker remains unexplained. 
* The need for a full variance decomposition is still required. 

In [93]:
#Outliers

In [94]:
ypw_p99_p01_ratio    = pwt_subset['GDP_pw'].quantile(0.99)/pwt_subset['GDP_pw'].quantile(0.01)
yhk_pw_p99_p01_ratio = pwt_subset['y_hk_pw'].quantile(0.99)/pwt_subset['y_hk_pw'].quantile(0.01)
ypw_p99_p01_ratio,yhk_pw_p99_p01_ratio

(8.363515913075576, 2.7908370897671846)

In [95]:
ypw_p95_p05_ratio    = pwt_subset['GDP_pw'].quantile(0.95)/pwt_subset['GDP_pw'].quantile(0.05)
yhk_pw_p95_p05_ratio = pwt_subset['y_hk_pw'].quantile(0.95)/pwt_subset['y_hk_pw'].quantile(0.05)
ypw_p95_p05_ratio,yhk_pw_p95_p05_ratio

(5.460361820923015, 2.0891265104682724)

In [96]:
ypw_p90_p10_ratio    = pwt_subset['GDP_pw'].quantile(0.90)/pwt_subset['GDP_pw'].quantile(0.10)
yhk_pw_p90_p10_ratio = pwt_subset['y_hk_pw'].quantile(0.90)/pwt_subset['y_hk_pw'].quantile(0.10)
ypw_p90_p10_ratio,yhk_pw_p90_p10_ratio

(3.3502616739254916, 1.8831095505822588)

In [97]:
ypw_p75_p25_ratio    = pwt_subset['GDP_pw'].quantile(0.75)/pwt_subset['GDP_pw'].quantile(0.25)
yhk_pw_p75_p25_ratio = pwt_subset['y_hk_pw'].quantile(0.75)/pwt_subset['y_hk_pw'].quantile(0.25)
ypw_p75_p25_ratio,yhk_pw_p75_p25_ratio

(2.189282284743125, 1.4108374090894757)

In [98]:
success1_pw_99_01 = yhk_pw_p99_p01_ratio/ypw_p99_p01_ratio
success1_pw_99_01

0.3336918490707923

In [99]:
success1_pw_95_05 = yhk_pw_p95_p05_ratio/ypw_p95_p05_ratio
success1_pw_95_05

0.38259854914067365

In [100]:
success1_pw_90_10 = yhk_pw_p90_p10_ratio/ypw_p90_p10_ratio
success1_pw_90_10

0.5620783490549933

In [101]:
success1_pw_75_25 = yhk_pw_p75_p25_ratio/ypw_p75_p25_ratio
success1_pw_75_25

0.6444291898406392

**Conclusion 2:**

* It seems that under the second measure proposed by Caselli, factors of production explain more than 100% of the variation in income per worker
* *This seems odd. If I didn't make any mistakes, it seems that outliers are big drivers of the results obtained in Conclusion 1.*
* Bottomline: if there is no mistake, we really need to think about outliers. An idea is to perform a variance decomposition but using samples without outliers and different regions of the world, and also samples over certain thresholds (i.e. over the median, over the25th percentilem etc)

**Conclusion 2 (SE without avh):**

* The code preceding conclusion 2 transforms actual GDP per capita and the factor-only GDP per capita using logs. Caselli takes the actual values, not the logs of the values. 
* With this in mind, the second measure proposed by Caselli yields fairly similar results to the first success measure:
    * The 90/10 and 95/5 yield success ratios between 24.8% and 36.1%. The average of these two numbers (30.5%) is very close to our first measure of success (28.1%). 
    * The 90/10 ratio in Caselli is 34%, broadly in line with our 90/10 ratio (36.1%). 
* Interestingly, the success ratio is consistently rising as we compare countries closer and closer to each other. The 75/25 yields a success ratio of 53.3% while the 99/1 ratio yields a success rate of 10.3% which is natural given that the countries are not too far from each other. 
* The >40 percentage point difference between the 99/1 and 75/25 success measures indicate that serious thought should be put into the treatment of outliers. All in all however, at least half of the cross-country income differences are pinned down by variation in technology and technology's interaction with factors of production. 

**Conclusion 2 (SE with avh):**

* The only alarm bells from proceeding with our project from including avh is that the 75/25 ratio indicates that just a third of the variation in income per capita is explained by TFP and its interaction with factors of production. 
    * As explained above, it can be argued that the 75/25 ratio compares countries sufficiently close to each other such that in many ways they are similar. According to the table produced by the next line of code, the difference in observed income per worker between the 25th and 75th percentiles is just over 1 standard deviation, whereas for the 95th and 5th percentiles it is around 3 standard deviations.
* Once countries further away from each other are compared, the success ratio falls to between 33.3% and 56.2%. 
* The big differences in success ratio when comparing outliers vs comparing countries closer together reiterate the need for careful consideration of treatment of outliers. 

In [113]:
pwt_subset['GDP_pw'].describe([0.05,0.10,0.25,0.50,0.75,0.90,0.95]).round(2)

count        61.00
mean      74177.78
std       36303.48
min       18371.85
5%        23559.62
10%       31758.38
25%       43949.43
50%       71907.51
75%       96217.70
90%      106398.89
95%      128644.07
max      219595.21
Name: GDP_pw, dtype: float64

## Disecting the Measure of Success

###  1. Shuting Down the Variation in Labour Shares

We start by recalculating the measures of success but shitting down the variation in labour shares. We do two alternative calculations:

 * **Alternative 1**: Fix the labour share at 2/3, the most widely used value for in the literature
 * **Alternative 2**: Fix the labour share at the median of the pwt sample in 2019: 0.56

In [114]:
pwt_subset['y_hk_pw_alt1'] = (pwt_subset['cn']**(1/3))*((pwt_subset['emp']*pwt_subset['avh']*pwt_subset['hc'])**(2/3))
pwt_subset['y_hk_pw_alt1'] = pwt_subset['y_hk_pw_alt1']/pwt_subset['emp']

In [ ]:
log_var_y_hk_pw_alt1 = np.log(np.var(pwt_subset['y_hk_pw_alt1']))
log_var_y_hk_pw_alt1

In [ ]:
success1_pw_alt1 = log_var_y_hk_pw_alt1/log_var_y_pw

In [ ]:
print("The Measure of Success Under *Alternative 1* is:") 
success1_pw_alt1.round(2)

In [ ]:
0.82/0.95

In [ ]:
pwt_subset['y_hk_pw_alt2'] = (pwt_subset['cn']**(1-0.56))*((pwt_subset['emp']*pwt_subset['avh']*pwt_subset['hc'])**(0.56))
pwt_subset['y_hk_pw_alt2'] = pwt_subset['y_hk_pw_alt2']/pwt_subset['emp']

In [ ]:
log_var_y_hk_pw_alt2 = np.log(np.var(pwt_subset['y_hk_pw_alt2']))
log_var_y_hk_pw_alt2.round(2)

In [ ]:
success1_pw_alt2 = log_var_y_hk_pw_alt2/log_var_y_pw

In [ ]:
print("The Measure of Success Under *Alternative 2* is:") 
success1_pw_alt2.round(2)

We choose **Alternative 1** to err on the conservative side. Shutting down the variability coming from differentials in factor intensities, we find that factors of production explain 81% of the variance in output per worker across countries.

### Variance Decomposition

To perform a full variance decomposition we need to calculate a few moments first.

In [ ]:
var_log_y_hk     = np.var(np.log(pwt_subset['y_hk_pw_alt1']))
var_log_y_hk

In [ ]:
var_log_k     = np.var(np.log(pwt_subset['k']))

In [ ]:
var_log_avh   = np.var(np.log(pwt_subset['avh']))

In [ ]:
var_log_h     = np.var(np.log(pwt_subset['hc']))

In [ ]:
cov_log_avh_h = np.cov(np.log(pwt_subset['avh']),np.log(pwt_subset['hc']))[0][1]

In [ ]:
cov_log_k_h   = np.cov(np.log(pwt_subset['k']),np.log(pwt_subset['hc']))[0][1]

In [ ]:
cov_log_k_avh = np.cov(np.log(pwt_subset['k']),np.log(pwt_subset['avh']))[0][1]

In [ ]:
α             = 1/3

In [ ]:
((α**2)*var_log_k)/var_log_y_hk

In [ ]:
(((1-α)**2)*var_log_avh)/var_log_y_hk

In [ ]:
(((1-α)**2)*var_log_h)/var_log_y_hk

In [ ]:
(2*((1-α)**2)*cov_log_avh_h)/var_log_y_hk

In [ ]:
(2*α*(1-α)*cov_log_k_h)/var_log_y_hk

In [ ]:
(2*α*(1-α)*cov_log_k_avh)/var_log_y_hk

In [ ]:
total_variation = np.abs(((α**2)*var_log_k)) + np.abs((((1-α)**2)*var_log_avh)) + np.abs((((1-α)**2)*var_log_h)) + np.abs((2*((1-α)**2)*cov_log_avh_h)) + np.abs((2*α*(1-α)*cov_log_k_h)) + np.abs((2*α*(1-α)*cov_log_k_avh))

In [ ]:
total_variation.round(2)

In [ ]:
#The share of the total variation given by capital per worker is

In [ ]:
(np.abs(((α**2)*var_log_k))/total_variation).round(2)

In [ ]:
#The share of the total variation given by hours worked is
(np.abs((((1-α)**2)*var_log_avh))/total_variation).round(2)

In [ ]:
#The share of the total variation given by hours worked is
(np.abs((((1-α)**2)*var_log_h))/total_variation).round(2)

In [ ]:
#The share of the total variation given by the interaction between hours worked and human capital
(np.abs((2*((1-α)**2)*cov_log_avh_h))/total_variation).round(2)

In [ ]:
#The share of the total variation given by the interaction between capital and human capital
(np.abs((2*α*(1-α)*cov_log_k_h))/total_variation).round(2)

In [ ]:
#The share of the total variation given by the interaction between capital and hours worked is
(np.abs((2*α*(1-α)*cov_log_k_avh))/total_variation).round(2)

## Conclusion

We perform a development accounting exercise ala Caselli, using PWT 10.0. We find that:

 * Factors of production explain 95% of the variation in income per worker
       * Out of that total variation, 86% is driven by pure factors of production, like capital, human capital, and hours worked, while the remaining 14% is explained by differences in factor shares across countries.
       * Within factors of production, differences in capital per worker explain 42% of the total variation, while human capital and hours account for 7% each.
       * We find that richer countries tend to have both higher levels of physical and human capital, with the joint comovement between these two factors of production explaining 21% of the share of income explained by factors only.
       * Moreover, rich countries tend to have higher levels of physical capital and lower levels of hours worked per worker, which reduces the total varition driven by factors of production, helping to explain 17% of the variation that is purely driven by productive factors
       * A remaining 5% of the total income differences explained by production factors is driven by the fact that in countries with higher levels of human capital workers tend to work fewer hours.

In [ ]:
#Suppose that we require our benchmark sample to have data on cgdpo, pop, cn, emp, and hc
#Now we want to see what the bias is due to sample selection in the following cases:
#1. Is the sample of countries that have data on cgdpo, pop, cn, emp, and hc (sample 1) different from the sample that has data on cgdpo, pop, cn, emp only (sample 2)?
#2. Is the sample of countries that have data on cgdpo, pop, cn, emp, hc, and labsh (sample 3) different from samples 1 and 2?
#3. Is the sample of countries that have data on cgdpo, pop, cn, emp, hc, labsh, and ctfp (sample 4) different from samples 1, 2, and 3?
#4. Is the sample of countries that have data on cgdpo, pop, cn, emp, hc, labsh, ctfp, and avh (sample 5) different from samples 1, 2, 3, and 4?

#Here we want to have descriptive statistics  for each sample on gdp per worker and per capita, capital per worker, and employment per capita
#We also want to do charts
#Then we compute the measures of success for each country
